File: src/day_6_part_1.rs
use std::fs;

fn transpose<T: Copy>(v: &Vec<Vec<T>>) -> Vec<Vec<T>> {
    let len = v[0].len();
    (0..len).map(|i| v.iter().map(|row| row[i]).collect()).collect()
}

pub fn main(scenario: &String, input_root: String) {
    let input_file = input_root + "/day_6_" + scenario + ".txt";
    let input = fs::read_to_string(input_file).expect("Failed to read input file");

    let mut vals: Vec<Vec<i64>> = Vec::new();
    let mut operators: Vec<&str> = Vec::new();
    for line in input.lines() {
        let line_chars = line.split_whitespace();
        let line_chars_copy = line_chars.clone();

        let line_vals = line_chars.map(|c| c.parse::<i64>().unwrap_or(0i64)).collect();
        vals.push(line_vals);
        operators = line_chars_copy.collect();
    }
    vals.pop();
    let vals_t = transpose(&vals);
    let mut results: Vec<i64> = Vec::with_capacity(vals.len());
    for (idx, op) in operators.iter().enumerate() {
        match *op {
            "+" => {
                results.push(vals_t[idx].iter().sum::<i64>());
            }
            "*" => {
                results.push(vals_t[idx].iter().product::<i64>());
            }
            _ => {}
        }
    }

    let grand_total:i64 = results.iter().sum();

    println!("The grand total is {}", grand_total);
}

File: python/day_6_part_1.py
import numpy as np
from core import read_file, parse_inputs


def solution(scenario: str):
    input_file = scenario + ".txt"
    input = read_file(input_file)

    lines = np.array([line.strip().split() for line in input.strip().splitlines()])
    numerical_data = np.array([[int(d) for d in line] for line in lines
                               if line[0] not in ["+", "*"]]).T

    result = 0
    for op, nums in zip(lines[-1], numerical_data):
        if op == "*":
            result += nums.prod()
        elif op == "+":
            result += nums.sum()

    print(f"grand total: {result}\n")


if __name__ == "__main__":
    scenario = parse_inputs("day_6_test_1")
    solution(scenario)
